harness/determined/common/api/certs.py

```diff
@@ -0,0 +1,220 @@
+
+import certifi
+import filelock
+
+from determined.common.api import authentication
+
+
+class Cert:
+    def __init__(
+        self,
+        cert_pem: Optional[str] = None,
+        noverify: bool = False,
+        name: Optional[str] = None,
+    ) -> None:
+        if cert_pem is not None and noverify:
+            raise AssertionError("you cannot set cert_pem with noverify=True")
+
+        if name == "":
+            name = None
+        self._name = name
+
+        if noverify:
+            self._bundle = False  # type: Union[None, str, bool]
+        elif cert_pem is None:
+            self._bundle = None
+        else:
+            # Don't use NamedTemporaryFile, since it would make the file inaccessible by path on
+            # Windows after this.
+            # (see https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile)
+            fd, combined_path = tempfile.mkstemp(prefix="det-master-cert-")
+            atexit.register(os.unlink, combined_path)
+
+            with open(fd, "wb") as out:
+                with open(certifi.where(), "rb") as base_certs:
+                    out.write(base_certs.read())
+                out.write(b"\n")
+                out.write(cert_pem.encode("utf8"))
+
+            self._bundle = combined_path
+
+    @property
+    def bundle(self) -> Union[None, str, bool]:
+        """
+        The path to a file containing an SSL certificate to trust specifically for the master, if
+        any, or False to disable cert verification entirely. If set to a path, it should always be a
+        temporary file that we own and can delete.
+        """
+        return self._bundle
+
+    @property
+    def name(self) -> Optional[str]:
+        """
+        The name we use to verify the master certificate.
+        """
+        return self._name
+
+
+cli_cert = None  # type: Optional[Cert]
+
+
+class _BrokenCertStore(Exception):
+    pass
+
+
+def _load_cert_store(path: pathlib.Path) -> Dict[str, str]:
+    if not path.exists():
+        return {}
+    try:
+
+        with path.open() as f:
+            content = f.read()
+
+        try:
+            store = json.loads(content)
+        except json.JSONDecodeError:
+            raise _BrokenCertStore()
+
+            if not isinstance(store, dict):
+                raise _BrokenCertStore()
+
+        # Store must be a dictionary.
+        if not isinstance(store, dict):
+            raise _BrokenCertStore()
+
+        # All keys are url's, all values are pem-encoded certs.
+        for k, v in store.items():
+            if not isinstance(k, str):
+                raise _BrokenCertStore()
+            if not isinstance(v, str):
+                raise _BrokenCertStore()
+
+    except _BrokenCertStore:
+        path.unlink()
+        return {}
+
+    return cast(Dict[str, str], store)
+
+
+@contextlib.contextmanager
+def _modifiable_store(path: pathlib.Path) -> Iterator[Dict["str", "str"]]:
```

Comment:
how does `persistent_store` sound to you. I think this function mainly persists changes to the cert store file.

Reply:
sounds fine